File: app/src/main/java/com/github/duchampdev/android/fot/frontend/TransactionDialogFragment.java
/*
 * Copyright (c) 2019 duchampdev.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.duchampdev.android.fot.frontend;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.appcompat.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;

import java.util.ArrayList;
import java.util.List;

import com.github.duchampdev.android.fot.R;
import com.github.duchampdev.android.fot.backend.SecondpartyAutocompleteService;
import com.github.duchampdev.android.fot.bdo.TransactionItem;
import com.github.duchampdev.android.fot.bdo.Category;
import com.github.duchampdev.android.fot.util.DateButton;
import com.github.duchampdev.android.fot.util.Util;


public class TransactionDialogFragment extends DialogFragment {

    private final List<TransactionDialogCallbacks> listeners = new ArrayList<>();

    private AutoCompleteTextView secondParty;
    private EditText amount;
    private EditText title;
    private DateButton date;
    private Spinner categorySpinner;

    private TransactionItem editable;
    private List<Category> categories;
    private ArrayAdapter<Category> categorySpinnerAdapter;
    private SecondpartyAutocompleteService autocompleteService;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        String dialogTitle = (args != null && args.containsKey("titem")) ? getResources().getString(R.string.tdialog_edit_transaction) : getResources().getString(R.string.tdialog_add_transaction);
        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(getActivity());
        dialogBuilder.setTitle(dialogTitle);
        dialogBuilder.setPositiveButton(getResources().getString(R.string.save), (dialog, which) -> {
            // implemented later
        });
        dialogBuilder.setNeutralButton(getResources().getString(R.string.abort), (dialog, which) -> {
            // nothing to do
        });


        View v = getActivity().getLayoutInflater().inflate(R.layout.dialog_transaction, null);

        secondParty = v.findViewById(R.id.tdialog_secondparty);
        amount = v.findViewById(R.id.tdialog_amount);
        title = v.findViewById(R.id.tdialog_title);
        date = v.findViewById(R.id.tdialog_date);
        categorySpinner = v.findViewById(R.id.tdialog_category);

        categorySpinnerAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_dropdown_item, new ArrayList<>());
        categorySpinner.setAdapter(categorySpinnerAdapter);

        amount.setOnFocusChangeListener((view, hasFocus) -> {
            if (!hasFocus) {
                // format currency correctly
                try {
                    double amount = Util.formatMoney(TransactionDialogFragment.this.amount.getText().toString());
                    TransactionDialogFragment.this.amount.setText(Util.formatMoney(amount));
                } catch (NumberFormatException nfe) {
                    Toast.makeText(getContext(), getResources().getString(R.string.tdialog_enter_valid_amount), Toast.LENGTH_SHORT).show();
                }
            }
        });
        autocompleteService = SecondpartyAutocompleteService.Companion.getInstance(secondParty.getContext());
        secondParty.setAdapter(new ArrayAdapter<>(secondParty.getContext(), android.R.layout.simple_spinner_dropdown_item,autocompleteService.getProposals()));
        dialogBuilder.setView(v);
        return dialogBuilder.create();
    }

    @Override
    public void onResume() { // set specific save routine - do not exit on error
        super.onResume();
        final AlertDialog dialog = (AlertDialog) getDialog();
        if (dialog != null) {
            dialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener(v -> saveTransaction());
        }
    }

    private void saveTransaction() {
        if (secondParty.getText().toString().isEmpty()) {
            getActivity().runOnUiThread(() -> Toast.makeText(getContext(), getResources().getString(R.string.tdialog_secondparty_missing), Toast.LENGTH_LONG).show());
        } else if (amount.getText().toString().isEmpty()) {
            getActivity().runOnUiThread(() -> Toast.makeText(getContext(), getResources().getString(R.string.tdialog_amount_missing), Toast.LENGTH_LONG).show());
        } else if (categorySpinner.getSelectedItemPosition() == AdapterView.INVALID_POSITION) {
            getActivity().runOnUiThread(() -> Toast.makeText(getContext(), getResources().getString(R.string.tdialog_category_missing), Toast.LENGTH_LONG).show());
        } else {
            // check amount format
            try {
                Util.formatMoney(amount.getText().toString());
            } catch (NumberFormatException nfe) {
                Toast.makeText(getContext(), getResources().getString(R.string.tdialog_enter_valid_amount), Toast.LENGTH_SHORT).show();
                return; // must break up, can't recover from this issue
            }

            // data is valid - persist
            if (editable == null) { // new item
                editable = new TransactionItem(secondParty.getText().toString(), Util.formatMoney(amount.getText().toString()), title.getText().toString(), (Category) categorySpinner.getSelectedItem(), date.getDate());
            } else {
                editable.setSecondParty(secondParty.getText().toString());
                editable.setAmount(Util.formatMoney(amount.getText().toString()));
                editable.setTitle(title.getText().toString());
                editable.setDate(date.getDate());
                editable.setCategory((Category) categorySpinner.getSelectedItem());
            }
            for (TransactionDialogCallbacks l : listeners) {
                l.dataSaved(editable);
            }
            dismiss(); // only called if valid input
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        Bundle args = getArguments();
        if (args != null && args.containsKey("categories")) {
            categories = (List<Category>) args.get("categories");
            categorySpinnerAdapter.clear();
            categorySpinnerAdapter.addAll(categories);
            categorySpinnerAdapter.notifyDataSetChanged();
            if (args.containsKey("titem")) {
                editable = (TransactionItem) args.getSerializable("titem");
                secondParty.setText(editable.getSecondParty());
                //amount.setText(editable.getAmount() + ""); // do not show euro sign
                amount.setText(Util.formatMoney(editable.getAmount()));
                title.setText(editable.getTitle());
                date.setDate(editable.getDate());
                categorySpinner.setSelection(categorySpinnerAdapter.getPosition(editable.getCategory()));
            }
        }
        return super.onCreateView(inflater, container, savedInstanceState);

    }

    public void registerTransactionDialogCallbacks(TransactionDialogCallbacks c) {
        listeners.add(c);
    }

    public interface TransactionDialogCallbacks {
        void dataSaved(TransactionItem item);
    }
}

File: app/src/main/java/com/github/duchampdev/android/fot/frontend/TransactionDialogFragment.kt
/*
 * Copyright (c) 2019 duchampdev.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.duchampdev.android.fot.frontend

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import com.github.duchampdev.android.fot.R
import com.github.duchampdev.android.fot.backend.FinanceOrgaToolDB
import com.github.duchampdev.android.fot.backend.SecondpartyAutocompleteService
import com.github.duchampdev.android.fot.bdo.Category
import com.github.duchampdev.android.fot.bdo.TransactionItem
import com.github.duchampdev.android.fot.util.NoFilterArrayAdapter
import com.github.duchampdev.android.fot.util.Util
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.android.synthetic.main.dialog_transaction.*
import kotlinx.android.synthetic.main.dialog_transaction.view.*

class TransactionDialogFragment: DialogFragment() {

    private val listeners: MutableList<TransactionDialogCallbacks> = ArrayList()

    private lateinit var editable: TransactionItem
    private lateinit var categories: List<Category>
    private lateinit var categorySpinnerAdapter: ArrayAdapter<Category>
    private lateinit var autocompleteService: SecondpartyAutocompleteService


    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialogTitle = if (arguments?.containsKey("titem") == true) resources.getString(R.string.tdialog_edit_transaction) else resources.getString(R.string.tdialog_add_transaction)
        
        val view = activity!!.layoutInflater.inflate(R.layout.dialog_transaction, null)
        
        return MaterialAlertDialogBuilder(activity)
                .setTitle(dialogTitle)
                .setPositiveButton(resources.getString(R.string.save)) { _, _ -> saveTransaction() } // callback implemented later
                .setNeutralButton(resources.getString(R.string.abort)) { _, _ -> } // callback implemented later
                .setView(view)
                .create()
                .also { onViewCreated(view, savedInstanceState) }
    }


    override fun onResume() { // set specific save routine - do not exit on error
        super.onResume()
        (dialog as AlertDialog?)
                ?.getButton(DialogInterface.BUTTON_POSITIVE)
                ?.setOnClickListener { saveTransaction() }
    }
    
    private fun saveTransaction() {
        val dbInstance = FinanceOrgaToolDB.getInstance(context!!)
        val selectedCategory =  dbInstance.getCategoryForName(dialog!!.tdialog_category.text.toString())
        when {
            dialog!!.tdialog_secondparty.text.toString().isEmpty() -> {
                activity!!.runOnUiThread { Toast.makeText(context, resources.getString(R.string.tdialog_secondparty_missing), Toast.LENGTH_LONG).show() }
            }
            dialog!!.tdialog_amount.text.toString().isEmpty() -> {
                activity!!.runOnUiThread { Toast.makeText(context, resources.getString(R.string.tdialog_amount_missing), Toast.LENGTH_LONG).show() }
            }
            selectedCategory == null -> {
                activity!!.runOnUiThread { Toast.makeText(context, resources.getString(R.string.tdialog_category_missing), Toast.LENGTH_LONG).show() }
            }
            else -> {
                // check amount format
                try {
                    Util.formatMoney(dialog!!.tdialog_amount.text.toString())
                } catch (e: NumberFormatException) {
                    Toast.makeText(context, resources.getString(R.string.tdialog_enter_valid_amount), Toast.LENGTH_SHORT).show()
                    return  // must break up, can't recover from this issue
                }

                // data is valid - persist
                if (! ::editable.isInitialized) { // new item
                    editable = TransactionItem(dialog!!.tdialog_secondparty.text.toString(), Util.formatMoney(dialog!!.tdialog_amount.text.toString()), dialog!!.tdialog_title.text.toString(), selectedCategory, dialog!!.tdialog_date.getDate())
                } else {
                    editable.secondParty = dialog!!.tdialog_secondparty.text.toString()
                    editable.amount = Util.formatMoney(dialog!!.tdialog_amount.text.toString())
                    editable.title = dialog!!.tdialog_title.text.toString()
                    editable.date = dialog!!.tdialog_date.getDate()
                    editable.category = selectedCategory
                }

                listeners.forEach { l -> l.dataSaved(editable) }
                dismiss() // only called if input was valid
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        categorySpinnerAdapter = NoFilterArrayAdapter(context!!, android.R.layout.simple_spinner_dropdown_item, ArrayList())
        view.tdialog_category.setAdapter(categorySpinnerAdapter)

        view.tdialog_amount.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
            if(!hasFocus) {
                // format currency correctly after field becomes unfocused
                try {
                    val amount = Util.formatMoney(view.tdialog_amount.text.toString())
                    view.tdialog_amount.setText(Util.formatMoney(amount))
                } catch (e: NumberFormatException) {
                    Toast.makeText(context, resources.getString(R.string.tdialog_enter_valid_amount), Toast.LENGTH_SHORT).show()
                }
            }
        }

        autocompleteService = SecondpartyAutocompleteService.getInstance(view.tdialog_secondparty.context!!)
        view.tdialog_secondparty.setAdapter(ArrayAdapter(view.tdialog_secondparty.context!!, android.R.layout.simple_spinner_dropdown_item, autocompleteService.getProposals()))

        if(arguments?.containsKey("categories") == true) {
            categories = arguments!!.get("categories")!! as List<Category>
            categorySpinnerAdapter.clear()
            categorySpinnerAdapter.addAll(categories)
            categorySpinnerAdapter.notifyDataSetChanged()

            if(arguments?.containsKey("titem") == true) {
                editable = arguments!!.get("titem")!! as TransactionItem

                view.tdialog_secondparty.setText(editable.secondParty)
                view.tdialog_amount.setText(Util.formatMoney(editable.amount))
                view.tdialog_title.setText(editable.title)
                view.tdialog_date.setDate(editable.date)
                view.tdialog_category.setText(editable.category.toString())
            }
        }
    }


    fun registerTransactionDialogCallbacks(c: TransactionDialogCallbacks) = listeners.add(c)

    interface TransactionDialogCallbacks {
        fun dataSaved(item: TransactionItem)
    }
}